x-pack/plugins/endpoint/server/routes/endpoints.ts

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { IRouter } from 'kibana/server';
+import { SearchResponse } from 'elasticsearch';
+import { schema } from '@kbn/config-schema';
+import { EndpointAppContext, EndpointData } from '../types';
+import { AllEndpointsQueryBuilder } from '../services/endpoint/endpoint_query_builders';
+
+interface HitSource {
+  _source: EndpointData;
+}
+
+export interface EndpointResultList {
+  endpoints: EndpointData[];
+  total: number;
+  requestPageSize: number;
+  requestIndex: number;
+}
+
+export function registerEndpointRoutes(router: IRouter, endpointAppContext: EndpointAppContext) {
+  router.get(
+    {
+      path: '/api/endpoint/endpoints',
+      validate: {
+        query: schema.object({
+          pageSize: schema.number({ defaultValue: 10, min: 1 }),
+          pageIndex: schema.number({ defaultValue: 0, min: 0 }),
+        }),
+      },
+      options: { authRequired: true },
+    },
+    async (context, req, res) => {
+      try {
+        const queryParams = await new AllEndpointsQueryBuilder(
```

Comment:
I'm not sure what the best way to _resolve_ this is (pun intended!), but I think if any one of `AllEndpointsQueryBuilder`, `.callAsCurrentUser`, or `mapToEndpointResultList` doesn't resolve, the `catch` won't fire (so neither res.ok or res.internalError will get called) and this will lock up. Also, errors thrown inside those functions won't make it out to this `catch` (I think). Maybe put it inside a Promise.race() with some kind of timeout?

Reply:
> (it does bubble through the await)
This is the expected behavior. 

There is no reason that I see why they will not resolve, the client call is a rest call that eventually timeouts and fails. The other reads the config to get the default paging, am not sure if this is an issue.